Create Python API for program_enrollments: Part I

kdmccormick opened the pull request:
@edx/masters-neem 

I'll be splitting this ticket into multiple parts in order to ease merge conflicts and code review.

This first part does some general refactoring and clean-up that will make the addition of a Python API easier. The single most significant change is making `test_views` more consistent and less dependent on very specific mocks. There are (hopefully) no actual functionality changes.

Next up will be renaming and reshuffling some files to make refactoring more manageable (WIP here: https://github.com/edx/edx-platform/pull/21556), then followed by the actual Python API.

https://openedx.atlassian.net/browse/EDUCATOR-4321

iloveagent57 commented on lms/djangoapps/program_enrollments/constants.py:
```diff
@@ -0,0 +1,33 @@
+    SUSPENDED = 'suspended'
+    CANCELED = 'canceled'
+    __ACTIVE__ = (ENROLLED, PENDING)
+    __ALL__ = (ENROLLED, PENDING, SUSPENDED, CANCELED)
```
Nit: I like how you made it clear that this variable has other dependencies, maybe consider a less generic name for it?  Perhaps `__MODEL_CHOICES__`?

kdmccormick replied:
I think I like the more generic name because there are non-model contexts where `__ALL__` is used (e.g., serializers). But explicit naming is also good.

How's this?
```python
...
__ALL__ = (ENROLLED, PENDING, SUSPENDED, CANCELED)

# Note: Any changes to this value will trigger a migration on ProgramEnrollment
__MODEL_CHOICES__ = __ALL__

iloveagent57 commented on lms/djangoapps/program_enrollments/management/commands/tests/test_link_program_enrollments.py:
```diff
@@ -4,24 +4,24 @@
+    COURSE_ENROLLMENT_ERR_TPL,
     DUPLICATE_KEY_TPL,
-    NO_PROGRAM_ENROLLMENT_TPL,
+    INCORRECT_PARAMETER_TPL,
```
Nit: while you're at, any appetite for getting rid of this `TPL` abbreviation and substituting `TEMPLATE`?  I can't help but to read `TPL` as "tuple".

iloveagent57 commented on lms/djangoapps/program_enrollments/models.py:
```diff
@@ -219,20 +215,23 @@ def enroll(self, user):
-            message = ("Attempted to create course enrollment for user={user} and course={course}"
-                       " but an enrollment already exists. Existing enrollment will be used instead")
-            logger.info(message.format(user=user.id, course=self.course_key))
+            message_fmt = (
```
Nit: Speaking of abbreviations :)

iloveagent57 commented on lms/djangoapps/program_enrollments/rest_api/v1/serializers.py:
```diff
@@ -19,72 +20,18 @@ class InvalidStatusMixin(object):
-                    return True
-        except KeyError:
-            pass
+        for status_error in self.errors.get('status', []):
```
haha, thank you for fixing this

iloveagent57 commented on lms/djangoapps/program_enrollments/rest_api/v1/serializers.py:
```diff
@@ -98,20 +45,9 @@ def get_account_exists(self, obj):
-class ProgramCourseEnrollmentListSerializer(serializers.Serializer):
-    """
-    Serializer for listing course enrollments in a program.
+    Serializer for display program-course enrollments.
```
"displaying"

iloveagent57 commented on lms/djangoapps/program_enrollments/rest_api/v1/serializers.py:
```diff
@@ -131,60 +67,57 @@ def get_curriculum_uuid(self, obj):
 
 
-class ProgramCourseGradeResult(object):
+class ProgramEnrollmentRequestMixin(serializers.Serializer, InvalidStatusMixin):
```
Nit: have the mixin class be the first thing to be inherited.

kdmccormick commented:
@iloveagent57 Addressed all your comments. This is what I settled on for `__MODEL_CHOICES__`: https://github.com/edx/edx-platform/pull/21511/files#diff-c2ac89648b258c4e185ab615a6bc604aR22; let me know how that looks

kdmccormick commented:
@iloveagent57 also, I'm reverting my changes to `[test-]link_program_enrollments.py` to avoid merge-conflict carnage with https://github.com/edx/edx-platform/pull/21541/

schenedx commented on openedx/core/djangoapps/catalog/utils.py:
```diff
@@ -573,6 +573,29 @@ def get_course_run_details(course_run_key, fields):
     return course_run_details
 
 
+def is_course_run_in_program(course_run_key, program):
```
Why do we need this function if we are not using this yet? Will be usage of this function come in later?

brittneyexline commented on lms/djangoapps/program_enrollments/rest_api/v1/utils.py:
```diff
@@ -1,22 +1,20 @@
 # -*- coding: utf-8 -*-
 """
-ProgramEnrollment internal api
```
hi @kdmccormick , the enterprise enrollment api is depending on these functions being exposed via the api.py file according to this ADR: https://github.com/edx/edx-platform/blob/master/docs/decisions/0002-inter-app-apis.rst

this file name change has resulted in breaking our functionality. there are a couple layers of weirdness here that resulted in this situation, but do you object in the mean time if i reintroduce the api.py file with functions wrapping the ones here?

i don't know how much future development is planned in this area, but it seems that these functions are not specific to the domain of program enrollments, and should be moved somewhere more appropriate since both programs and enterprise portal enrollment functionality depend on these functions. let me know what you think.

fyi @christopappas

kdmccormick replied:
Hi @brittneyexline. The purpose of this PR is to work towards introducing a Python API as described in that ADR. This `api.py` file didn't meet the requirements of the ADR, and the "internal utilities" wording in the docstring made me think it didn't have any external dependents, hence me renaming it.

For the time being I have no qualms with an `api.py` here that is essentially:
```python
"""
docstring about why enterprise API needs this module to exist,
and link to ticket that will remove it
"""
from .utils import (
    # all the functions you need
)
```

In terms of development in this area, I'm currently in the process of tidying up the program_enrollments app. [This PR](https://github.com/edx/edx-platform/pull/21591) will start introducing a program_enrollments Python API that can be depended upon. Like you said, I agree that the functions that were in `api.py` and are now in `utils.py` belong somewhere else. Maybe the `courseware` app?

brittneyexline replied:
hi @kdmccormick - some of the weirdness i was referring to is the fact that we were importing these functions from edx-enterprise, so the dependencies were not readily visible working only in edx-platform. in the mean time, i reintroduced the imports we were depending on (https://github.com/edx/edx-platform/pull/21593/files#diff-26734a293ffb45f242140771adde9de7R3) - if you could preserve those for the time being, we would appreciate it!

i have filed a ticket for us to follow up on moving these utility functions somewhere else: https://openedx.atlassian.net/browse/ENT-2294. we should be able to address this in the next 2-3 weeks - if it becomes more urgent let us know and we can reprioritize.

kdmccormick replied:
@brittneyexline Cool, can you look at https://github.com/edx/edx-platform/pull/21605/files? Thanks for the ticket!
